src/dns/local.rs

```diff
@@ -1,134 +1,166 @@
+#[derive(Default)]
 pub struct LocalOverrideTable {
     /// Entries in the local override table. First iteration of the resolver: a simple list that
     /// will be queried O(n). Later iterations will use a more efficient data structure like a tree
     /// per domain part (e.g. com, org, net, etc.)
     entries: Vec<DNSEntry>,
-    round_robin_ptrs: HashMap<String, usize>,
 }
 
-impl Default for LocalOverrideTable {
-    fn default() -> LocalOverrideTable {
-        LocalOverrideTable {
+impl LocalOverrideTable {
+
+    pub fn new() -> LocalOverrideTable {
+        let mut table = LocalOverrideTable {
             entries: Vec::new(),
-            round_robin_ptrs: HashMap::new(),
-        }
+        };
+
+        table.reload_table_entries();
+        table
     }
-}
 
-impl LocalOverrideTable {
+    /// Regenerates the new entries table
+    pub fn reload_table_entries(&mut self) {
+        // @todo: this should reload all table entries from the configuration into the self.entries list
+    }
 
-    pub fn resolve(&mut self, domain: &str) -> Option<IpAddr> {
-        let mut ips = Vec::new();
-        let mut wildcard_ips = Vec::new();
+    pub fn resolve(&mut self, domain: &str, resolve_type: ResolveType) -> Result<IpAddr, types::Error> {
+        let mut matched_entry = None;
+        let mut matched_wildcard_entries = Vec::new();
 
         // For now it's ok to iterate the entries list, but later we should use a more efficient
-        // data structure.
-        for entry in &self.entries {
-            // We separate wildcard matches and specific matches.
+        // data structure in the future
+        for (idx, entry) in self.entries.iter().enumerate() {
+            // We separate wildcard matches and specific matches (e.g. *.example.com vs example.com)
             if WildMatch::new(&entry.domain_name).matches(domain) {
-                wildcard_ips.push(entry.ip);
+                matched_wildcard_entries.push(idx);
             }
             if entry.domain_name == domain {
-                ips.push(entry.ip);
+                // First match is ok to return
+                matched_entry = Some(idx);
+                break;
             }
         }
 
         // if we haven't found a specific domain match, we use the wildcard domain, otherwise
         // the specific matches take precedence.
-        if ips.len() == 0 {
-            ips = wildcard_ips
+        if matched_entry.is_none() {
+            // Entry most specific match
+            matched_entry = Some(self.find_most_specific_match(matched_wildcard_entries))
         }
 
-        // If there is only one ip for a domain, return it
-        if ips.len() == 1 {
-            return Some(*ips.get(0).unwrap());
-        }
+        // Fetch the actual entry in the entries list
+        let found_entry = self.entries.get_mut(matched_entry.expect("No idx found")).expect("No entry found");
```

Comment:
Here again, don't use expect, I mean this function already returns a Result. Maybe it would be nice to implement the functionality for the `?` syntax. If this is not already done, i can do this if you want.